ksqldb-execution/src/main/java/io/confluent/ksql/execution/interpreter/CastInterpreter.java

```diff
@@ -0,0 +1,253 @@
+    } else if (from.baseType() == SqlBaseType.BOOLEAN) {
+      return object -> object;
+    }
+    throw new KsqlException("Unsupported cast to BOOLEAN: " + from);
+  }
+
+  public static CastFunction castToArrayFunction(
+      final SqlType from,
+      final SqlType to,
+      final KsqlConfig config
+  ) {
+    if (from.baseType() == SqlBaseType.ARRAY
+        && to.baseType() == SqlBaseType.ARRAY) {
+      final SqlArray fromArray = (SqlArray) from;
+      final SqlArray toArray = (SqlArray) to;
+      final CastFunction itemCastFunction = castFunction(
+          fromArray.getItemType(), toArray.getItemType(), config);
+      return o ->
+          CastEvaluator.castArray((List<?>) o, itemCastFunction::cast);
+    }
+    throw new KsqlException(getErrorMessage(ConversionType.CAST, from, to));
+  }
+
+  public static CastFunction castToMapFunction(
+      final SqlType from,
+      final SqlType to,
+      final KsqlConfig config
+  ) {
+    if (from.baseType() == SqlBaseType.MAP
+        && to.baseType() == SqlBaseType.MAP) {
+      final SqlMap fromMap = (SqlMap) from;
+      final SqlMap toMap = (SqlMap) to;
+      final CastFunction keyCastFunction = castFunction(
+          fromMap.getKeyType(), toMap.getKeyType(), config);
+      final CastFunction valueCastFunction = castFunction(
+          fromMap.getValueType(), toMap.getValueType(), config);
+      return o ->
+          CastEvaluator.castMap((Map<?, ?>) o, keyCastFunction::cast, valueCastFunction::cast);
+    }
+    throw new KsqlException("Unsupported cast to " + to + ": " + from);
+  }
+
+  /**
+   * Conversion functions that work for converting type either during comparison or casting.
+   */
+  public static class NumberConversions {
+    public static Double toDouble(final Object object, final SqlType from,
+        final ConversionType type) {
+      if (object == null) {
+        return null;
+      }
+      if (object instanceof Number) {
+        return ((Number) object).doubleValue();
+      } else {
+        throw new KsqlException(getErrorMessage(type, from, SqlTypes.DOUBLE));
+      }
+    }
+
+    public static Long toLong(final Object object, final SqlType from, final ConversionType type) {
+      if (object == null) {
+        return null;
+      }
+      if (object instanceof Number) {
+        return ((Number) object).longValue();
+      } else {
+        throw new KsqlException(getErrorMessage(type, from, SqlTypes.BIGINT));
+      }
+    }
+
+    public static Integer toInteger(final Object object, final SqlType from,
+        final ConversionType type) {
+      if (object == null) {
+        return null;
+      }
+      if (object instanceof Number) {
+        return ((Number) object).intValue();
+      } else {
+        throw new KsqlException(getErrorMessage(type, from, SqlTypes.INTEGER));
+      }
+    }
+  }
+
+  private static String getErrorMessage(
+      final ConversionType type,
+      final SqlType from,
+      final SqlType to
+  ) {
```

Comment:
do we have any coding style agreed on for newlines? I saw some places we have a newline for ) { after argument list and some places we do not.

Reply:
I don't know if we have a fully consistent approach, but I'll try to make it all consistent with either all on one line or one line per arg.